pkg/server/plugin/nodeattestor/httpchallenge/httpchallenge_test.go

```diff
@@ -0,0 +1,423 @@
+			hclConf:     "allow_non_root_ports = false",
+			tofu:        true,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      1024,
+			}),
+		},
+		{
+			name:        "Attest fails if hostname is not valid by dns pattern",
+			expErr:      "rpc error: code = PermissionDenied desc = nodeattestor(http_challenge): the requested hostname is not allowed to connect",
+			hclConf:     `allowed_dns_patterns = ["p[0-9][.]example[.]com"]`,
+			tofu:        true,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      80,
+			}),
+		},
+		{
+			name:        "Attest fails if nonce does not match",
+			expErr:      "rpc error: code = PermissionDenied desc = nodeattestor(http_challenge): challenge verification failed: expected nonce \"YmFkMTIzNDU2Nzg5YWJjZGVmZ2hpamtsbW5vcHFyc3Q=\"",
+			hclConf:     "",
+			tofu:        true,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      80,
+			}),
+		},
+		{
+			name:        "Attest fails when reattesting with tofu",
+			expErr:      "rpc error: code = PermissionDenied desc = nodeattestor(http_challenge): attestation data has already been used to attest an agent",
+			hclConf:     "",
+			tofu:        false,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      80,
+			}),
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.tofu {
+				common_httpchallenge.DefaultRandReader = strings.NewReader("bad123456789abcdefghijklmnopqrstuvwxyz")
+			} else {
+				common_httpchallenge.DefaultRandReader = strings.NewReader("123456789abcdefghijklmnopqrstuvwxyz")
+			}
+			plugin := loadPlugin(t, tt.hclConf, !tt.tofu)
+			result, err := plugin.Attest(context.Background(), tt.payload, tt.challengeFn)
+			require.Contains(t, err.Error(), tt.expErr)
+			require.Nil(t, result)
+		})
+	}
+}
+
+func TestAttestSucceeds(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	    if r.URL.Path != "/.well-known/spiffe/nodeattestor/http_challenge/default/challenge" {
+	        t.Errorf("Expected to request '/.well-known/spiffe/nodeattestor/http_challenge/default/challenge', got: %s", r.URL.Path)
+	    }
+	    w.WriteHeader(http.StatusOK)
+	    w.Write([]byte(`MTIzNDU2Nzg5YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnc=`))
+	}))
+	defer server.Close()
+
+	u, _ := url.Parse(server.URL)
+	_, port, _ := net.SplitHostPort(u.Host)
+	oldDialContext := http.DefaultTransport.(*http.Transport).DialContext
+	dialContext := func(ctx context.Context, network, addr string) (net.Conn, error) {
```

Comment:
Why do we need this special dialing? Can the hostname be "localhost" instead of "foo"?

Reply:
chatted via slack about this... Maybe resolved now. Please rereview.